src/coverage/renderer.ts

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 Espressif Systems (Shanghai) CO LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as vscode from "vscode";
+import {
+  generateCoverageForEditors,
+  textEditorWithCoverage,
+} from "./coverageService";
+import * as idfConf from "../idfConfiguration";
+
+export interface CoverageOptions {
+  darkThemeCoveredBackgroundColor: string;
+  lightThemeCoveredBackgroundColor: string;
+  darkThemePartialBackgroundColor: string;
+  lightThemePartialBackgroundColor: string;
+  darkThemeUncoveredBackgroundColor: string;
+  lightThemeUncoveredBackgroundColor: string;
+  overviewRuleLane: vscode.OverviewRulerLane;
+}
+
+export function getCoverageOptions() {
+  const coverageOptions: CoverageOptions = {
+    darkThemeCoveredBackgroundColor: idfConf.readParameter(
+      "idf.coveredDarkTheme"
+    ),
+    darkThemeUncoveredBackgroundColor: idfConf.readParameter(
+      "idf.uncoveredDarkTheme"
+    ),
+    darkThemePartialBackgroundColor: idfConf.readParameter(
+      "idf.partialDarkTheme"
+    ),
+    lightThemeCoveredBackgroundColor: idfConf.readParameter(
+      "idf.coveredLightTheme"
+    ),
+    lightThemePartialBackgroundColor: idfConf.readParameter(
+      "idf.partialLightTheme"
+    ),
+    lightThemeUncoveredBackgroundColor: idfConf.readParameter(
+      "idf.uncoveredLightTheme"
+    ),
+    overviewRuleLane: vscode.OverviewRulerLane.Right,
+  };
+  return coverageOptions;
+}
+
+export class CoverageRenderer {
+  private coveredDecoratorType: vscode.TextEditorDecorationType;
+  private partialDecoratorType: vscode.TextEditorDecorationType;
+  private notCoveredDecoratorType: vscode.TextEditorDecorationType;
+  private countDecoratorTypes: vscode.TextEditorDecorationType[];
+  private marginDecoratorType: vscode.TextEditorDecorationType;
+  private workspaceFolder: vscode.Uri;
```

Comment:
Color scheme is bit confusing, maybe add some tooltip if possible near the left hand corner text showing line numbers

Also by default coverage color we should add bright light green for covered lines since number is already denoting the hit count!

<img width="654" alt="Screen Shot 2020-05-19 at 3 22 35 PM" src="https://user-images.githubusercontent.com/25585652/82312467-b797e600-99e4-11ea-9dee-69b05d424609.png">

Reply:
I made an initial implementation that had partial coverage for line count=1 and full coverage for count > 1 and uncovered for count = 0. We should probably do as gcovr does and implement covered for >90%*MaxLineCount, 90%*MaxLineCount>partial>75%*MaxLineCount and 75%*MaxLineCount<low covered.

Reply:
This could be cumbersome, maybe we can just show covered and un-covered with green and red simply and show the hit count as showing now on left hand side